design/service-discovery.md

```diff
@@ -0,0 +1,178 @@
+
+On the server, each time a new request is received, and already knowing which is the GUID of the response subscriber that corresponds to the publisher of the request, it checks if such subscriber has already been matched. If so, the request is processed as usual. Else, the request is held back until the subscriber is matched.
+Whenever a response subscriber is matched, the list of pending requests is checked looking for requests corresponding to that client, and sent to processing.
+
+The solution is quite symmetrical the other way round. The server creates the response publisher before the request subscriber. When creating the request subscriber, it will insert the response publisher's GUID on the `USER_DATA` of the subscriber, so that the client can know both remote endoints belong to the same server.
+
+On the client, whenever a request subscriber is matched, it checks if the corresponding response publisher has already been matched. If so, the server's <publisher_guid, subscriber_guid> is added to a list of fully-matched servers. Else, it is added to a list of half-matched servers.
+Whenever a response publisher is matched, the list of half-matched servers is checked, and if it is found there, it is moved to the fully-matched list.
+
+#### Caveats ####
+
+* Un-matching of endpoints has to deal with the new internal structures to maintain coherence. For example, removing the server guids from the fully-matched list and possibly moving it to the half-matched list, or cleaning the list of pending requests.
+* The algorithm has to be able support remote endpoints that do not add the response GUID on the `USER_DATA`. This is to keep compatibility with older versions and other vendors. In these cases, legacy behavior is acceptable.
+
+#### Note on the GUID format on the USER_DATA ####
+
+At the moment, the `USER_DATA` is not being used on the RMW endpoints. We can simply add the GUID in text form using `operator<<` and `operator>>` to read and write.
+
+In order be able to add other information on the `USER_DATA` on the future, we must *tag* the information somehow. The proposed format is:
+
+```
+  responseGUID:<GUID>
+```
+
+where `responseGUID:` is a string literal and `<GUID>` is the char string form of the GUID of the related endpoint, as formatted by `operator<<`.
+
+Using `properties` instead of `USER_DATA` would be preferable in this case, but unfortunately `properties` are not available at publisher/subscriber level on Fast DDS at this moment.
+
+## Class diagram ##
+
+The following class diagram only shows classes modified by this solution. And for those classes, only added members (`+`), modified (`~`) or removed (`-`) members are shown.
+
+![Class diagram](service-discovery-images/class-diagram.svg)
```

Comment:
I'm guessing you mean the rich diff on the 'files changed' view of the pull request is not able to resolve the images. I'd say the problem is with how Github manages the pull requests. We have no trouble with the images on our branch, they are rendered correctly and in their place. I tried a pull request on our own fork, and it looks everything gets sorted out once the merge is done.

Although Github cannot display the SVGs integrated on the markdown rich diff, it can render the images themselves when selecting rich diff on the image file. The name of the image and the alternative text displayed on the markdown are the same (save for name convention adaptations).

Reply:
Good to know.